Added support for additional sensors.  Tweaked the device_info method

jakecolman opened the pull request:
First pass at some simple changes to add support for additional devices.

andyalm commented:
I'm wondering if a `binary_sensor` is the right kind of entity to be using for some of these. A binary_sensor primarily consists of something that is either "on" or "off" (determined by the `is_on` method). While you can certainly include additional `state_attributes` as metadata along for the ride, I think there are other entities that would be better fits for temperature and humidity sensors, specifically. Take a look at the HA documentation of the various entity types they support here: https://developers.home-assistant.io/docs/core/entity. The `climate` and/or `humidifier` sensors look like they could be relevent. If neither of those, then a `sensor` is always a good fallback (unlike a `binary_sensor`, supports arbitrary values).

For the motion detector, I think a `binary_sensor` probably is a good fit. `is_on` would be based on whether motion is currently being detected or not. However, the `ScoutDoorWindowSensor` class will probably not work as-is for that. For example, I think the motion detector would need to have its own `is_on` implementation. It could make sense to extract a common base class for these two. They could both be bootstrapped in the `async_setup_entry` method at the top of the `binary_sensor.py` file.

jakecolman commented:
All of the sensors I added - motion, water, smoke - I think fit the profile
of a binary sensor in that they have either been activated or not.  The
device classes that I used to define the new sensors support that as well.
So I think the changes in my first commit make sense as they are.  My
second commit is simply adding temperature and humidity as metadata since
the data is available.  I agree however, that if we want to use these Scout
sensors to actually report and track temperature and humidity, we should
track the physical device as two different logical sensors.  Are we both
saying the same thing or do you have some concerns with either or both
commits?


On Mon, Feb 8, 2021 at 4:34 PM Andy Alm <notifications@github.com> wrote:

> I'm wondering if a binary_sensor is the right kind of entity to be using
> for some of these. A binary_sensor primarily consists of something that is
> either "on" or "off" (determined by the is_on method). While you can
> certainly include additional state_attributes as metadata along for the
> ride, I think there are other entities that would be better fits for
> temperature and humidity sensors, specifically. Take a look at the HA
> documentation of the various entity types they support here:
> https://developers.home-assistant.io/docs/core/entity. The climate and/or
> humidifier sensors look like they could be relevent. If neither of those,
> then a sensor is always a good fallback (unlike a binary_sensor, supports
> arbitrary values).
>
> For the motion detector, I think a binary_sensor probably is a good fit.
> is_on would be based on whether motion is currently being detected or
> not. However, the ScoutDoorWindowSensor class will probably not work
> as-is for that. For example, I think the motion detector would need to have
> its own is_on implementation. It could make sense to extract a common
> base class for these two. They could both be bootstrapped in the
> async_setup_entry method at the top of the binary_sensor.py file.
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/andyalm/homeassistant-scout-alarm/pull/1#issuecomment-775477941>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ADL7RIT3BLOTJD3G6RAVKSDS6BKFTANCNFSM4XJL5XUA>
> .
>

jakecolman commented:
I just read the link and you are correct.  Temperature and humidity should
not be included in the device_state_attributes.  I will remove it in a
later commit.  Might not be a bad idea to leave it in place for now just as
a data point for when we build out the climate and/or humidifier sensor for
this same device.  The question is can you create two logical sensors from
the same physical one?

On Mon, Feb 8, 2021 at 5:03 PM Jake Colman <jake.colman@gmail.com> wrote:

> All of the sensors I added - motion, water, smoke - I think fit the
> profile of a binary sensor in that they have either been activated or not.
> The device classes that I used to define the new sensors support that as
> well.  So I think the changes in my first commit make sense as they are.
> My second commit is simply adding temperature and humidity as metadata
> since the data is available.  I agree however, that if we want to use these
> Scout sensors to actually report and track temperature and humidity, we
> should track the physical device as two different logical sensors.  Are we
> both saying the same thing or do you have some concerns with either or both
> commits?
>
>
> On Mon, Feb 8, 2021 at 4:34 PM Andy Alm <notifications@github.com> wrote:
>
>> I'm wondering if a binary_sensor is the right kind of entity to be using
>> for some of these. A binary_sensor primarily consists of something that is
>> either "on" or "off" (determined by the is_on method). While you can
>> certainly include additional state_attributes as metadata along for the
>> ride, I think there are other entities that would be better fits for
>> temperature and humidity sensors, specifically. Take a look at the HA
>> documentation of the various entity types they support here:
>> https://developers.home-assistant.io/docs/core/entity. The climate
>> and/or humidifier sensors look like they could be relevent. If neither
>> of those, then a sensor is always a good fallback (unlike a binary_sensor,
>> supports arbitrary values).
>>
>> For the motion detector, I think a binary_sensor probably is a good fit.
>> is_on would be based on whether motion is currently being detected or
>> not. However, the ScoutDoorWindowSensor class will probably not work
>> as-is for that. For example, I think the motion detector would need to have
>> its own is_on implementation. It could make sense to extract a common
>> base class for these two. They could both be bootstrapped in the
>> async_setup_entry method at the top of the binary_sensor.py file.
>>
>> —
>> You are receiving this because you authored the thread.
>> Reply to this email directly, view it on GitHub
>> <https://github.com/andyalm/homeassistant-scout-alarm/pull/1#issuecomment-775477941>,
>> or unsubscribe
>> <https://github.com/notifications/unsubscribe-auth/ADL7RIT3BLOTJD3G6RAVKSDS6BKFTANCNFSM4XJL5XUA>
>> .
>>
>

jakecolman commented:
I updated the open Pull Request with one more commit to remove temp and
humidity from the device state.  I think that's it for now from me with
binary sensors.  After you have a chance to review it, let me know if you
have any concerns with my PR or if you plan on accepting it.  Next, when I
have time, I am going to look at doing a sensors implementation for temp
and humidity.  Modeling on Humidity and Climate sensors does not seem to be
appropriate since those are intended for devices that can effect change.
The Scout sensors are read-only sensors that simply report a value so
generic Sensors seem to be the way to go.  I am going to model the
implementation on the Xiaomi code which supports multiple logical sensors
from one device that can report multiple values (i.e., temp, humidity,
etc).  Happy to hear any thoughts you may have.  Just drop me an email or a
LinkedIn message.  Thanks for allowing me to help with this.

On Mon, Feb 8, 2021 at 5:23 PM Jake Colman <jake.colman@gmail.com> wrote:

> I just read the link and you are correct.  Temperature and humidity should
> not be included in the device_state_attributes.  I will remove it in a
> later commit.  Might not be a bad idea to leave it in place for now just as
> a data point for when we build out the climate and/or humidifier sensor for
> this same device.  The question is can you create two logical sensors from
> the same physical one?
>
> On Mon, Feb 8, 2021 at 5:03 PM Jake Colman <jake.colman@gmail.com> wrote:
>
>> All of the sensors I added - motion, water, smoke - I think fit the
>> profile of a binary sensor in that they have either been activated or not.
>> The device classes that I used to define the new sensors support that as
>> well.  So I think the changes in my first commit make sense as they are.
>> My second commit is simply adding temperature and humidity as metadata
>> since the data is available.  I agree however, that if we want to use these
>> Scout sensors to actually report and track temperature and humidity, we
>> should track the physical device as two different logical sensors.  Are we
>> both saying the same thing or do you have some concerns with either or both
>> commits?
>>
>>
>> On Mon, Feb 8, 2021 at 4:34 PM Andy Alm <notifications@github.com> wrote:
>>
>>> I'm wondering if a binary_sensor is the right kind of entity to be
>>> using for some of these. A binary_sensor primarily consists of something
>>> that is either "on" or "off" (determined by the is_on method). While
>>> you can certainly include additional state_attributes as metadata along
>>> for the ride, I think there are other entities that would be better fits
>>> for temperature and humidity sensors, specifically. Take a look at the HA
>>> documentation of the various entity types they support here:
>>> https://developers.home-assistant.io/docs/core/entity. The climate
>>> and/or humidifier sensors look like they could be relevent. If neither
>>> of those, then a sensor is always a good fallback (unlike a
>>> binary_sensor, supports arbitrary values).
>>>
>>> For the motion detector, I think a binary_sensor probably is a good
>>> fit. is_on would be based on whether motion is currently being detected
>>> or not. However, the ScoutDoorWindowSensor class will probably not work
>>> as-is for that. For example, I think the motion detector would need to have
>>> its own is_on implementation. It could make sense to extract a common
>>> base class for these two. They could both be bootstrapped in the
>>> async_setup_entry method at the top of the binary_sensor.py file.
>>>
>>> —
>>> You are receiving this because you authored the thread.
>>> Reply to this email directly, view it on GitHub
>>> <https://github.com/andyalm/homeassistant-scout-alarm/pull/1#issuecomment-775477941>,
>>> or unsubscribe
>>> <https://github.com/notifications/unsubscribe-auth/ADL7RIT3BLOTJD3G6RAVKSDS6BKFTANCNFSM4XJL5XUA>
>>> .
>>>
>>

jakecolman commented:
I think that that support for Scount sensors as temp/humidity sensors is looking pretty good.  We may want to change the 'name' method to NOT include a T/H prefix since the icon already makes it clear that it is a Temp/Humidity sensor.  I did it this way to easily distinguish between the multiple instances of the same sensor but I'm not sure we should keep it this way.  Let me know what you think.

andyalm commented on custom_components/scout_alarm/sensor.py:
```diff
@@ -0,0 +1,145 @@
+    devices = await location_api.get_devices()
+
+    for d in devices:
+        if d['type'] in SUPPORTED_SCOUT_DEVICE_TYPES_TEMPERATURE:
```
Instead of basing this logic on the device type, can we do it by the presence of a `temperature` attribute under `reported` attribute in the device payload? e.g. 
```python
if d['reported'].get('temperature'):
```

Similar for `humidity`. My door panels are an older model and actually don't support temp or humidity readings.

andyalm commented on custom_components/scout_alarm/binary_sensor.py:
```diff
@@ -94,17 +109,19 @@ def device_state_attributes(self):
+            "manufacturer": self._device['reported']['manufacturer'],
+            "name": self.name,
+            "sw_version": self._device['reported']['fw_version'],
+            "model": self._device['reported']['model']
```
My devices don't have a `model` attribute. Since python hashes throw when a key doesn't exist via `[]`, can we change this to `.get('model')` instead?

jakecolman commented:
Done.  Please let me know if you find any other issues.

Here is a screenshot of how the temp/humidity sensors are displayed

[image: image.png]

On Mon, Feb 15, 2021 at 1:05 AM Andy Alm <notifications@github.com> wrote:

> *@andyalm* commented on this pull request.
> ------------------------------
>
> In custom_components/scout_alarm/binary_sensor.py
> <https://github.com/andyalm/homeassistant-scout-alarm/pull/1#discussion_r575959343>
> :
>
> >          }
>
>      @property
>      def device_info(self):
>          """Return device registry information for this entity."""
>          return {
>              "identifiers": {(DOMAIN, self._device['id'])},
> -            "manufacturer": "Scout Alarm",
> -            "name": self.name
> +            "manufacturer": self._device['reported']['manufacturer'],
> +            "name": self.name,
> +            "sw_version": self._device['reported']['fw_version'],
> +            "model": self._device['reported']['model']
>
> My devices don't have a model attribute. Since python hashes throw when a
> key doesn't exist via [], can we change this to .get('model') instead?
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/andyalm/homeassistant-scout-alarm/pull/1#pullrequestreview-590179464>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ADL7RIU3TY5E3F6FGD36V4LS7C2Q3ANCNFSM4XJL5XUA>
> .
>

andyalm commented on custom_components/scout_alarm/sensor.py:
```diff
@@ -0,0 +1,131 @@
+
+    @property
+    def should_poll(self) -> bool:
+        return  True
```
Its worth noting that this integration is categorized as a "cloud push" integration in home assistant, which means generally speaking it does not poll, but relies on a real time push API. This component breaks with that right now. That said, in the context of a temperature and humidity sensor, polling probably works fine, so I think that's ok for now. At some point we should review the pusher events that fire to see if any of them could be used instead of polling.

andyalm commented:
The `sensor` support is looking good to me. I'm concerned about the `binary_sensor` changes as-written, as it will start creating binary sensors for several new device types, but I'm not confident that the `is_on` method (the core method of a `binary_sensor`) will work for these new device types. Have you verified that part?

jakecolman commented:
With respect to "push" versus "poll", sensors such as those used for temp
and humidity have to be polled since we need to track changes as they
happen.  There is no push event that tells you the temp or humidity changed
so you have no choice but to poll.  HA handles automatic polling of sensors
every few minutes but, unless force_update is true, the data is not updated
unless the value actually changes.  I tested using the force_update mode
and confirmed that it does poll correctly.

I just ran some tests on the new water detector sensor and I was NOT able
to get it to trigger. Silly of me to think that it would just work in the
same way that the access_sensors do.  Maybe I should just update the Pull
Request to remove the new sensors from binary_sensors until we figure this
out?  This way we still have the temp/humidity sensor support in Master
while I continue to figure out how to support the newer binary sensors.
What do you think?

On Mon, Feb 15, 2021 at 8:01 PM Andy Alm <notifications@github.com> wrote:

> The sensor support is looking good to me. I'm concerned about the
> binary_sensor changes as-written, as it will start creating binary
> sensors for several new device types, but I'm not confident that the is_on
> method (the core method of a binary_sensor) will work for these new
> device types. Have you verified that part?
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/andyalm/homeassistant-scout-alarm/pull/1#issuecomment-779518203>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ADL7RIU4PLGQCE44JAFACBLS7G7VNANCNFSM4XJL5XUA>
> .
>

andyalm commented on custom_components/scout_alarm/binary_sensor.py:
```diff
@@ -70,13 +82,38 @@ def is_on(self):
+            if co_state is None:
+                co_state = "ok"
+            """on_state is true if either smoke or co are not ok"""
+            on_state = (smoke_state != SCOUT_DEVICE_STATE_OK or co_state != SCOUT_DEVICE_STATE_OK)
```
I think ideally we create separate entities for smoke and co sensors.  Home assistant supports creating multiple entities for the same physical device, and that would help keep the logic simpler here.
